src/api/async_resource.cc

```diff
@@ -18,36 +19,55 @@ AsyncResource::AsyncResource(Isolate* isolate,
     : env_(Environment::GetCurrent(isolate)),
       resource_(isolate, resource) {
   CHECK_NOT_NULL(env_);
-  async_context_ = EmitAsyncInit(isolate, resource, name,
-                                 trigger_async_id);
+  env_->SetAsyncResourceContextFrame(
+      reinterpret_cast<std::uintptr_t>(this),
+      {isolate, async_context_frame::current(isolate)});
+  async_context_ = EmitAsyncInit(isolate, resource, name, trigger_async_id);
 }
 
 AsyncResource::~AsyncResource() {
+  CHECK_NOT_NULL(env_);
+  env_->RemoveAsyncResourceContextFrame(reinterpret_cast<std::uintptr_t>(this));
```

Comment:
Should we change the sequence here?
In most places you enter context frame before asyncId handling and exit/restore after. But here it's the other way around.
I guess this has impact to anyone using a destroy hook for whatever reason.

Reply:
Timing consistency between ALS and async_hooks was never deterministic. Promises also produce a similar timing inconsistency. I can move that around if you feel strongly about it, but I don't think it really matters much.

Reply:
Did a fast test and it seems destroy hook is not running with the the store as of now. So actually this sequence seems to be less breaking one.

In general I think that changing the implementation details of ALS should not result in too many side effects to other node APIs.
I know AsyncHooks API is experimental and their use for ALS will/should go away. But there are other async hooks users which might also rely on ALS therefore the result of `als.getStore()` seen within init/before/after/destroy hooks should stay the same if possible.
I expect there are no tests for this because till now async hooks were fully bound to ALS and quite a lot testing happens on async hooks level as of now.

Reply:
Yeah, so my long-term plan is to:

- migrate ALS off async_hooks
- create some _new_ API for resource tracking that doesn't conflate consumable and non-consumable resource, plus probably have some filtering mechanism to subscribe to specific resource types and skip the barrier hop altogether when not interested.
- eventually deprecate async_hooks

This will be a very long-term thing as I only have so much free time to work on this stuff, but that's my general aim with this. 🙂